crml/nft/src/benchmarking.rs

```diff
@@ -0,0 +1,274 @@
+	verify {
+		assert!(<Nft<T>>::listings(&collection_id, token_id).is_some());
+	}
+
+	direct_purchase {
+		let owner: T::AccountId = account("owner", 0, 0);
+		let buyer: T::AccountId = account("buyer", 0, 0);
+		let collection_id = setup_token::<T>(owner.clone());
+		let token_id = T::TokenId::from(0_u32);
+		let payment_asset = 16_000;
+		let price = 1_000_000 * 10_000; // 1 million 4dp asset
+		let _ = T::MultiCurrency::deposit_creating(&buyer, Some(payment_asset), price);
+		let _ = <Nft<T>>::direct_sale(RawOrigin::Signed(owner.clone()).into(), collection_id.clone(), token_id, Some(buyer.clone()), payment_asset, price).expect("listed ok");
+
+	}: _(RawOrigin::Signed(buyer.clone()), collection_id.clone(), token_id)
+	verify {
+		assert_eq!(<Nft<T>>::token_owner(&collection_id, token_id), buyer);
+	}
+
+	auction {
+		let owner: T::AccountId = account("owner", 0, 0);
+		let collection_id = setup_token::<T>(owner.clone());
+		let token_id = T::TokenId::from(0_u32);
+		let payment_asset = 16_000;
+		let reserve_price = 1_000_000 * 10_000; // 1 million 4dp asset
+		let duration = T::BlockNumber::from(100_u32);
+
+	}: _(RawOrigin::Signed(owner.clone()), collection_id.clone(), token_id, payment_asset, reserve_price, Some(duration))
+	verify {
+		assert!(<Nft<T>>::listings(&collection_id, token_id).is_some());
+	}
+
+	bid {
+		let owner: T::AccountId = account("owner", 0, 0);
+		let buyer: T::AccountId = account("buyer", 0, 0);
+
+		let collection_id = setup_token::<T>(owner.clone());
+		let token_id = T::TokenId::from(0_u32);
+		let payment_asset = 16_000;
+		let reserve_price = 1_000_000 * 10_000; // 1 million 4dp asset
+		let duration = T::BlockNumber::from(100_u32);
+
+		let _ = T::MultiCurrency::deposit_creating(&owner, Some(payment_asset), reserve_price);
+		let _ = T::MultiCurrency::deposit_creating(&buyer, Some(payment_asset), reserve_price + 1);
+		let _ = <Nft<T>>::auction(RawOrigin::Signed(owner.clone()).into(), collection_id.clone(), token_id, payment_asset, reserve_price, Some(duration)).expect("listed ok");
+		// worst case path is to replace an existing bid
+		let _ = <Nft<T>>::bid(RawOrigin::Signed(owner.clone()).into(), collection_id.clone(), token_id, reserve_price);
+
+	}: _(RawOrigin::Signed(buyer.clone()), collection_id.clone(), token_id, reserve_price + 1)
+	verify {
+		assert_eq!(<Nft<T>>::listing_winning_bid(&collection_id, token_id), Some((buyer.clone(), reserve_price + 1)));
+	}
+
+	cancel_sale {
+		let owner: T::AccountId = account("owner", 0, 0);
+		let collection_id = setup_token::<T>(owner.clone());
+		let token_id = T::TokenId::from(0_u32);
+		let payment_asset = 16_000;
+		let reserve_price = 1_000_000 * 10_000; // 1 million 4dp asset
+		let duration = T::BlockNumber::from(100_u32);
+
+		let _ = <Nft<T>>::auction(RawOrigin::Signed(owner.clone()).into(), collection_id.clone(), token_id, payment_asset, reserve_price, Some(duration)).expect("listed ok");
```

Comment:
Again here according to the following code:
`			<CollectedTokens<T>>::mutate(&collection_id, current_owner, |tokens| {
				tokens.retain(|t| t != &token_id)
			});
`
It seesm we should have a number of auctions scheduled to make it either the worst or average case scenario.